geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-parquet/src/main/scala/org/locationtech/geomesa/fs/storage/parquet/FilterConverter.scala

```diff
@@ -23,14 +23,18 @@ import scala.reflect.ClassTag
 
 object FilterConverter {
 
-  def convert(sft: SimpleFeatureType, filter: Filter): (Option[FilterPredicate], Option[Filter]) = {
-    if (filter == Filter.INCLUDE) { (None, None) } else {
-      FilterHelper.propertyNames(filter).foldLeft((Option.empty[FilterPredicate], Option(filter)))(reduce(sft))
+  def convert(sft: SimpleFeatureType, filter: Filter): (Int => (Option[FilterPredicate], Option[Filter])) = {
+    def thunk(version: Int): (Option[FilterPredicate], Option[Filter]) = {
+      if (filter == Filter.INCLUDE) { (None, None) } else {
+        FilterHelper.propertyNames(filter).foldLeft((Option.empty[FilterPredicate], Option(filter)))(reduce(sft, version))
+      }
     }
+    thunk
```

Comment:
this is a good place to use memoization - we're going to be calling this repeatedly for a given query, and the only input that changes is (maybe) the version. so we can do something like:
```
if (filter == Filter.INCLUDE) { _ => (None, None) } else {
  val propertyNames = FilterHelper.propertyNames(filter)
  lazy val v1 = propertyNames.foldLeft((Option.empty[FilterPredicate], Option(filter)))(reduce(sft, 1))
  lazy val v2 = propertyNames.foldLeft((Option.empty[FilterPredicate], Option(filter)))(reduce(sft, 2))
  (i) => if (i == 2) { v2 } else { v1 }
}
```